ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -1501,6 +1523,10 @@ public RelOptMaterialization apply(RelOptMaterialization materialization) {
           // Add view-based rewriting rules to planner
           planner.addRule(HiveMaterializedViewRule.INSTANCE_PROJECT_FILTER);
           planner.addRule(HiveMaterializedViewRule.INSTANCE_FILTER);
+          planner.addRule(HiveMaterializedViewRule.INSTANCE_PROJECT_JOIN);
```

Comment:
Adding so many rules will slow down compilation. All these rules slightly vary in pattern they are looking for and roughly does same thing once they apply.
I suspect we have to create so many rules because Calcite's pattern matching on operator tree isnt powerful. We shall try to improve on this otherwise we traverse tree multiple times which could be avoided.

Reply:
Yes, I guess that will depend on how the planner (Volcano in this case) handles internally the triggering of rules, we should aim at triggering them doing a single pass over the query plan.
Although currently Calcite has no way to indicate that a rule should not be triggered only once over a specific set of equivalent expressions, I think we should be safe on that end because we have only the MV rewriting specific rules in this group, thus they will only work over current plan.